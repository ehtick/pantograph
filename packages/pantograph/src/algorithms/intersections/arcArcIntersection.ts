import { Vector } from "../../definitions.js";
import { Arc } from "../../models/segments/Arc.js";
import removeDuplicatePoints from "../../utils/removeDuplicatePoints.js";
import {
  add,
  distance,
  normalize,
  perpendicular,
  sameVector,
  scalarMultiply,
  subtract,
} from "../../vectorOperations";

const complementArc = (arc: Arc): Arc => {
  const { firstPoint, lastPoint, center, clockwise } = arc;
  return new Arc(lastPoint, firstPoint, center, clockwise, {
    ignoreChecks: true,
  });
};

const handleOverlaps = (arc1: Arc, arc2: Arc): Arc[] => {
  // handle the case with common points at start or end first
  if (arc1.isSame(arc2)) {
    return [arc1];
  }
  // two arcs with common points at start and end (but not the same arc)

  const points = removeDuplicatePoints(
    [
      arc2.isOnSegment(arc1.firstPoint) ? arc1.firstPoint : null,
      arc2.isOnSegment(arc1.lastPoint) ? arc1.lastPoint : null,
      arc1.isOnSegment(arc2.firstPoint) ? arc2.firstPoint : null,
      arc1.isOnSegment(arc2.lastPoint) ? arc2.lastPoint : null,
    ].filter((p) => p !== null) as Vector[],
    // We sort by the param value of the first arc. This means that the points
    // will be sorted with the same orientation than arc1
  ).sort((a, b) => arc1.pointToParam(a) - arc1.pointToParam(b));

  if (points.length === 0) return [];
  // We consider the case when the arcs touch only on
  // the last point. We consider that they do not overlap there
  //
  // We might want to revisit this choice
  else if (points.length === 1) return [];
  else if (points.length === 2) {
    // Similar to the case with length 1, we ignore the double overlapping point
    if (arc1.isSame(complementArc(arc2))) return [];
    return [new Arc(points[0], points[1], arc1.center, arc1.clockwise)];
  } else if (points.length === 3) {
    // Similar to the case with length 1, we ignore the single overlapping point
    const startIndex =
      sameVector(points[0], arc2.lastPoint) ||
      sameVector(points[0], arc2.firstPoint)
        ? 1
        : 0;
    return [
      new Arc(
        points[0 + startIndex],
        points[1 + startIndex],
        arc1.center,
        arc1.clockwise,
      ),
    ];
  } else if (points.length === 4) {
    return [
      new Arc(points[0], points[1], arc1.center, arc1.clockwise),
      new Arc(points[2], points[3], arc1.center, arc1.clockwise),
    ];
  }
  throw new Error("Bug in the arc arc overlap algorithm");
};

export function arcArcIntersection(
  arc1: Arc,
  arc2: Arc,
  includeOverlaps = false,
  precision?: number,
): Vector[] | Arc[] {
  const epsilon = precision ? precision : arc1.precision;
  const centersDistance = distance(arc1.center, arc2.center);

  const radiusSum = arc1.radius + arc2.radius;

  // The circles do not touch
  if (centersDistance > radiusSum + epsilon) {
    return [];
  }

  const radiusDifference = Math.abs(arc1.radius - arc2.radius);

  // The arcs are concentric
  if (centersDistance < radiusDifference - epsilon) {
    return [];
  }

  // With a common center we can have overlaps
  if (centersDistance < epsilon) {
    if (radiusDifference > epsilon) {
      return [];
    } else {
      if (!includeOverlaps) {
        return [];
      }
      return handleOverlaps(arc1, arc2);
    }
  }

  const centersVector = normalize(subtract(arc2.center, arc1.center));
  // The circles are tangent to each other
  const isOutsideTangent = centersDistance > radiusSum - epsilon;
  if (
    // circles are outside each other
    isOutsideTangent ||
    // circles are inside each other
    Math.abs(centersDistance - radiusDifference) < epsilon
  ) {
    const orientation = isOutsideTangent || arc1.radius > arc2.radius ? 1 : -1;
    const intersectionPoint = add(
      arc1.center,
      scalarMultiply(centersVector, orientation * arc1.radius),
    );

    if (
      arc1.isOnSegment(intersectionPoint) &&
      arc2.isOnSegment(intersectionPoint)
    ) {
      return [intersectionPoint];
    } else {
      return [];
    }
  }

  // The circles cross each other
  const radiusToChord =
    (arc1.radius * arc1.radius) / (2 * centersDistance) -
    (arc2.radius * arc2.radius) / (2 * centersDistance) +
    centersDistance / 2;

  const midPoint = add(
    arc1.center,
    scalarMultiply(centersVector, radiusToChord),
  );

  const halfChord = Math.sqrt(
    arc1.radius * arc1.radius - radiusToChord * radiusToChord,
  );

  const chordVector = perpendicular(centersVector);

  const p1 = add(midPoint, scalarMultiply(chordVector, halfChord));
  const p2 = add(midPoint, scalarMultiply(chordVector, -halfChord));

  const intersections = [];
  if (arc1.isOnSegment(p1) && arc2.isOnSegment(p1)) {
    intersections.push(p1);
  }
  if (arc1.isOnSegment(p2) && arc2.isOnSegment(p2)) {
    intersections.push(p2);
  }

  return intersections;
}
