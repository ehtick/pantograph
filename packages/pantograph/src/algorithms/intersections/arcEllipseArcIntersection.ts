import type { EllipseArc } from "../../models/segments/EllipseArc";
import type { Arc } from "../../models/segments/Arc";
import { ellipseEllipseIntersection } from "./ellipseEllipseIntersection";

export function arcEllipseArcIntersection(arc: Arc, ellipseArc: EllipseArc) {
  const points = ellipseEllipseIntersection(arc, ellipseArc);
  return points.filter((p) => arc.isOnSegment(p) && ellipseArc.isOnSegment(p));
}
