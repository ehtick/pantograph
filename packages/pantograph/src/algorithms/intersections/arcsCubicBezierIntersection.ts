import { EllipseArc } from "../../models/segments/EllipseArc.js";
import { Arc } from "../../models/segments/Arc.js";
import { Vector } from "../../definitions.js";
import removeDuplicateValues from "../../utils/removeDuplicateValues.js";
import { solveGenericPolynomial } from "../solvers/solvePolynomials.js";
import { CubicBezier } from "../../models/segments/CubicBezier.js";

/*
 * We use wolfram alpha to do the analitical resolution of the intersection
 * Collect[ Eliminate[{a*x^2+b*x*y+c*y^2+d * x + e * y + f == 0, p3 * t^3 + p2 * t^2  + p1 * t +p0 == x,q3*t^3+q2*t^2+q1*t+q0==y}, {x,y}], t ]
 * */

const intersectionsPolynomial = (arc: EllipseArc | Arc, curve: CubicBezier) => {
  const [[p0, p1, p2, p3], [q0, q1, q2, q3]] = curve.polynomialCoefficients;

  const el1 = arc.coefficients;
  const a = el1.x2;
  const b = el1.xy;
  const c = el1.y2;
  const d = el1.x;
  const e = el1.y;
  const f = el1.c;

  const p02 = p0 * p0;
  const p12 = p1 * p1;
  const p22 = p2 * p2;
  const p32 = p3 * p3;
  const q02 = q0 * q0;
  const q12 = q1 * q1;
  const q22 = q2 * q2;
  const q32 = q3 * q3;

  const z0 = f + d * p0 + a * p02 + e * q0 + b * p0 * q0 + c * q02;
  const z1 =
    d * p1 +
    2 * a * p0 * p1 +
    b * p1 * q0 +
    e * q1 +
    b * p0 * q1 +
    2 * c * q0 * q1;
  const z2 =
    a * p12 +
    d * p2 +
    2 * a * p0 * p2 +
    b * p2 * q0 +
    b * p1 * q1 +
    c * q12 +
    e * q2 +
    b * p0 * q2 +
    2 * c * q0 * q2;
  const z3 =
    2 * a * p1 * p2 +
    d * p3 +
    2 * a * p0 * p3 +
    b * p3 * q0 +
    b * p2 * q1 +
    b * p1 * q2 +
    2 * c * q1 * q2 +
    e * q3 +
    b * p0 * q3 +
    2 * c * q0 * q3;
  const z4 =
    a * p22 +
    2 * a * p1 * p3 +
    b * p3 * q1 +
    b * p2 * q2 +
    c * q22 +
    b * p1 * q3 +
    2 * c * q1 * q3;
  const z5 = 2 * a * p2 * p3 + b * p3 * q2 + b * p2 * q3 + 2 * c * q2 * q3;
  const z6 = a * p32 + b * p3 * q3 + c * q32;

  return [z0, z1, z2, z3, z4, z5, z6];
};

export function arcsCubicBezierIntersection(
  arc: EllipseArc | Arc,
  curve: CubicBezier
): Vector[] {
  const epsilon = Math.max(arc.precision, curve.precision);

  const polynomial = intersectionsPolynomial(arc, curve);
  const solutions = solveGenericPolynomial(polynomial, epsilon).filter((t) => {
    return t >= -curve.precision && t <= 1 + curve.precision;
  });

  return removeDuplicateValues(solutions, epsilon)
    .map((t) => {
      return curve.paramPoint(t);
    })
    .filter((p) => {
      return arc.isOnSegment(p);
    });
}
