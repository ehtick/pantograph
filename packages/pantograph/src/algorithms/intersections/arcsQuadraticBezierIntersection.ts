import { EllipseArc } from "../../models/segments/EllipseArc.js";
import { Arc } from "../../models/segments/Arc.js";
import { Vector } from "../../definitions.js";
import removeDuplicateValues from "../../utils/removeDuplicateValues.js";
import { solveQuartic } from "../solvers/solvePolynomials.js";
import { QuadraticBezier } from "../../models/segments/QuadraticBezier.js";

/*
 * We use wolfram alpha to do the analitical resolution of the intersection
 * Collect[ Eliminate[{a*x^2+b*x*y+c*y^2+d * x + e * y + f == 0, p2 * t^2  + p1 * t +p0 == x,q2*t^2+q1*t+q0==y}, {x,y}], t ]
 * */

const intersectionsPolynomial = (
  arc: EllipseArc | Arc,
  curve: QuadraticBezier
): [number, number, number, number, number] => {
  const [[p0, p1, p2], [q0, q1, q2]] = curve.polynomialCoefficients;

  const el1 = arc.coefficients;
  const a = el1.x2;
  const b = el1.xy;
  const c = el1.y2;
  const d = el1.x;
  const e = el1.y;
  const f = el1.c;

  const p02 = p0 * p0;
  const p12 = p1 * p1;
  const p22 = p2 * p2;
  const q02 = q0 * q0;
  const q12 = q1 * q1;
  const q22 = q2 * q2;

  const z0 = a * p02 + b * p0 * q0 + c * q02 + d * p0 + e * q0 + f;
  const z1 =
    2 * a * p0 * p1 +
    b * p0 * q1 +
    b * p1 * q0 +
    2 * c * q0 * q1 +
    d * p1 +
    e * q1;
  const z2 =
    2 * a * p0 * p2 +
    a * p12 +
    b * p0 * q2 +
    b * p1 * q1 +
    b * p2 * q0 +
    2 * c * q0 * q2 +
    c * q12 +
    d * p2 +
    e * q2;

  const z3 = 2 * a * p1 * p2 + b * p1 * q2 + b * p2 * q1 + 2 * c * q1 * q2;
  const z4 = a * p22 + b * p2 * q2 + c * q22;

  return [z0, z1, z2, z3, z4];
};

export function arcsQuadraticBezierIntersection(
  arc: EllipseArc | Arc,
  curve: QuadraticBezier
): Vector[] {
  const epsilon = Math.max(arc.precision, curve.precision);

  const polynomial = intersectionsPolynomial(arc, curve);
  const solutions = solveQuartic(...polynomial).filter((t) => {
    return t >= -curve.precision && t <= 1 + curve.precision;
  });

  return removeDuplicateValues(solutions, epsilon)
    .map((t) => {
      return curve.paramPoint(t);
    })
    .filter((p) => {
      return arc.isOnSegment(p);
    });
}
