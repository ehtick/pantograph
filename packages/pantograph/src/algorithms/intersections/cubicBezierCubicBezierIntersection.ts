import { Vector } from "../../definitions.js";
import { CubicBezier } from "../../models/segments/CubicBezier.js";
import { sameVector } from "../../vectorOperations.js";
import { bezierClip } from "./bezierClip.js";

export function handleOverlaps(curve1: CubicBezier, curve2: CubicBezier) {
  const commonPoints: Vector[] = [];
  const toTest: [Vector, CubicBezier][] = [
    [curve1.firstPoint, curve2],
    [curve1.lastPoint, curve2],
    [curve2.firstPoint, curve1],
    [curve2.lastPoint, curve1],
  ];

  // This could be optimised by doing only one vector => parameter conversion

  toTest.forEach(([point, curve]) => {
    if (curve.isOnSegment(point)) {
      commonPoints.push(point);
    }
  });

  if (commonPoints.length < 2) {
    // no overlap
    return null;
  }

  if (commonPoints.length === 2) {
    return [curve1.splitAt(commonPoints)[1]];
  }

  if (commonPoints.length === 3) {
    // there is one curve that is within the other
    if (
      sameVector(commonPoints[0], curve1.firstPoint) &&
      sameVector(commonPoints[1], curve1.lastPoint)
    ) {
      return [curve1];
    }
    return [curve2];
  }

  if (commonPoints.length === 4) {
    return [curve1];
  }
}

export function cubicBezierCubicBezierIntersection(
  curve1: CubicBezier,
  curve2: CubicBezier,
  includeOverlaps = false
) {
  const epsilon = Math.max(curve1.precision, curve2.precision);

  if (includeOverlaps) {
    const overlappingCurve = handleOverlaps(curve1, curve2);
    if (overlappingCurve) {
      return overlappingCurve;
    }
  }

  return bezierClip(curve1, curve2, epsilon);
}
