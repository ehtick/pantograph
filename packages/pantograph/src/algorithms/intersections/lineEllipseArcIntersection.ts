import { Line } from "../../models/segments/Line.js";
import { EllipseArc } from "../../models/segments/EllipseArc.js";
import { Vector } from "../../definitions.js";

// Using the derivation from http://www.nabla.hr/CS-EllipseAndLine1.htm
export function lineEllipseArcIntersection(
  line: Line,
  arc: EllipseArc,
  precision = 1e-9,
) {
  const lineP = line.transform(arc.ellipseReferenceFrameTransform);

  const m = lineP.slope;
  const c = lineP.yIntercept;

  const a2 = arc.majorRadius * arc.majorRadius;
  const b2 = arc.minorRadius * arc.minorRadius;
  const ab = arc.majorRadius * arc.minorRadius;

  const m2 = lineP.slope * lineP.slope;
  const c2 = lineP.yIntercept * lineP.yIntercept;

  // Helper function to convert back the intersection from the ellipse
  // reference frame to the original frame
  // It also checks that the intersection is within the bounds of the segments
  const filterIntersectionInRef = (intersections: Vector[]) =>
    intersections
      .map((point) =>
        arc.reverseEllipseReferenceFrameTransform.transform(point),
      )
      .filter((point) => line.isOnSegment(point) && arc.isOnSegment(point));

  // We handle the case of the line being vertical
  if (!Number.isFinite(m)) {
    // Vertical line
    const x = lineP.firstPoint[0];

    // Outside
    if (Math.abs(x) - arc.majorRadius > precision) return [];

    // Tangent
    if (Math.abs(Math.abs(x) - arc.majorRadius) < precision) {
      return filterIntersectionInRef([[x, 0]]);
    }

    const y = arc.minorRadius * Math.sqrt(1 - (x * x) / a2);

    const p1: Vector = [x, y];
    const p2: Vector = [x, -y];

    return filterIntersectionInRef([p1, p2]);
  }

  const discriminant = a2 * m2 + b2 - c2;

  if (discriminant < -precision) {
    return [];
  }

  const denominator = a2 * m2 + b2;

  // We have a tangent line
  if (Math.abs(discriminant) < precision) {
    const x = -(a2 * m * c) / denominator;
    const y = (b2 * c) / denominator;
    return filterIntersectionInRef([[x, y]]);
  }

  const sqrtDiscriminant = Math.sqrt(discriminant);

  const p1: Vector = [
    -(a2 * m * c + ab * sqrtDiscriminant) / denominator,
    (b2 * c - ab * m * sqrtDiscriminant) / denominator,
  ];
  const p2: Vector = [
    -(a2 * m * c - ab * sqrtDiscriminant) / denominator,
    (b2 * c + ab * m * sqrtDiscriminant) / denominator,
  ];

  return filterIntersectionInRef([p1, p2]);
}
