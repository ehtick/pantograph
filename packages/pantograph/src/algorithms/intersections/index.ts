import { Line } from "../../models/segments/Line";

import type { Vector } from "../../definitions";
import { Segment } from "../../models/segments/Segment";
import { Arc } from "../../models/segments/Arc";
import { lineArcIntersection } from "./lineArcIntersection";
import { lineLineIntersection } from "./lineLineIntersection";
import { arcArcIntersection } from "./arcArcIntersection";

export function findIntersections(
  segment1: Segment,
  segment2: Segment,
  precision?: number
): Vector[] {
  if (segment1 instanceof Line && segment2 instanceof Line) {
    const intersection = lineLineIntersection(
      segment1,
      segment2,
      false,
      precision
    );
    if (intersection === null) return [];
    return [intersection as Vector];
  }
  if (segment1 instanceof Line && segment2 instanceof Arc) {
    return lineArcIntersection(segment1, segment2, precision);
  }
  if (segment1 instanceof Arc && segment2 instanceof Line) {
    return lineArcIntersection(segment2, segment1, precision);
  }
  if (segment1 instanceof Arc && segment2 instanceof Arc) {
    return arcArcIntersection(segment1, segment2, false, precision) as Vector[];
  }

  throw new Error("Not implemented");
}

export function findIntersectionsAndOverlaps(
  segment1: Segment,
  segment2: Segment,
  precision?: number
): { intersections: Vector[]; overlaps: Segment[]; count: number } {
  // If we have two lines, checks are fast enough to not use bounding boxes
  if (segment1 instanceof Line && segment2 instanceof Line) {
    const intersection = lineLineIntersection(
      segment1,
      segment2,
      true,
      precision
    );
    if (intersection === null)
      return { intersections: [], overlaps: [], count: 0 };
    if (intersection instanceof Line)
      return { intersections: [], overlaps: [intersection], count: 1 };
    return { intersections: [intersection], overlaps: [], count: 1 };
  }

  if (!segment1.boundingBox.overlaps(segment2.boundingBox)) {
    return { intersections: [], overlaps: [], count: 0 };
  }

  if (segment1 instanceof Line && segment2 instanceof Arc) {
    const intersections = lineArcIntersection(segment1, segment2, precision);
    return { intersections, overlaps: [], count: intersections.length };
  }

  if (segment1 instanceof Arc && segment2 instanceof Line) {
    const intersections = lineArcIntersection(segment2, segment1, precision);
    return { intersections, overlaps: [], count: intersections.length };
  }

  if (segment1 instanceof Arc && segment2 instanceof Arc) {
    const intersections = arcArcIntersection(
      segment1,
      segment2,
      true,
      precision
    );
    if (!intersections.length)
      return { intersections: [], overlaps: [], count: 0 };
    if (intersections[0] instanceof Arc)
      return {
        intersections: [],
        overlaps: intersections as Arc[],
        count: intersections.length,
      };
    return {
      intersections: intersections as Vector[],
      overlaps: [],
      count: intersections.length,
    };
  }

  throw new Error("Not implemented");
}
