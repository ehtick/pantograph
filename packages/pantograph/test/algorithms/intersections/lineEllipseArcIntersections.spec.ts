import { describe, it, expect } from "vitest";

import { lineEllipseArcIntersection } from "../../../src/algorithms/intersections/lineEllipseArcIntersection";
import { Line } from "../../../src/models/segments/Line";
import { EllipseArc } from "../../../src/models/segments/EllipseArc";

//import { debugImg, dpnt } from "../../debug";

describe("lineEllipseArcIntersections", () => {
  it("should return the intersection point of a line and an arc", () => {
    const line = new Line([-1, -1], [1, 1]);

    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false);
    const intersections = lineEllipseArcIntersection(line, arc);

    //debugImg([line, arc, ...intersections.map((p) => dpnt(p))]);
    expect(intersections.length).toBe(1);
    expect(intersections[0]).toBeVector([0, 0]);
  });

  it("should return multiple intersections of a line and an arc", () => {
    const line = new Line([-5, -5], [1, 1]);
    const arc = new EllipseArc(
      [0, 0],
      [2, 1],
      [2, 0],
      2,
      1,
      0,
      false,
    ).translate(-3.2, -2);
    const intersections = lineEllipseArcIntersection(line, arc);

    //debugImg([line, arc, ...intersections.map((p) => dpnt(p))]);
    expect(intersections.length).toBe(2);
    expect(intersections).toMatchSnapshot();
  });

  it("should return nothing when crossing outisde of the arcs", () => {
    const line = new Line([-1, 1], [2, 0]);

    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false);
    const intersections = lineEllipseArcIntersection(line, arc);

    //debugImg([line, arc, ...intersections.map((p) => dpnt(p))]);
    expect(intersections.length).toBe(0);
  });

  it("should handle tangent lines to the arc", () => {
    const line = new Line([4, -1], [4, 1]).rotate(33);
    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false).rotate(
      33,
    );

    const intersections = lineEllipseArcIntersection(line, arc);

    //debugImg([line, arc, ...intersections.map((p) => dpnt(p))]);
    expect(intersections.length).toBe(1);
    expect(intersections).toMatchSnapshot();
  });

  it("should handle vertical lines with the arc", () => {
    const line = new Line([3.5, -1], [3.5, 1]).rotate(33);
    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false).rotate(
      33,
    );

    const intersections = lineEllipseArcIntersection(line, arc);

    //debugImg([line, arc, ...intersections.map((p) => dpnt(p))]);
    expect(intersections.length).toBe(2);
    expect(intersections).toMatchSnapshot();
  });

  it("should return nothing when there is not intersection", () => {
    const line = new Line([12, -1], [12, 1]);
    const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false).rotate(
      33,
    );

    const intersections = lineEllipseArcIntersection(line, arc);
    expect(intersections.length).toBe(0);
  });
});
