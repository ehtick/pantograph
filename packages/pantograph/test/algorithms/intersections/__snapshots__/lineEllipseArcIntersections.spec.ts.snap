// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lineEllipseArcIntersections > should handle tangent lines to the arc 1`] = `
[
  [
    3.354682271781696,
    2.178556140060108,
  ],
]
`;

exports[`lineEllipseArcIntersections > should handle vertical lines with the arc 1`] = `
[
  [
    2.575102127571993,
    2.460965061225817,
  ],
  [
    3.2955918480459747,
    1.3515081838793719,
  ],
]
`;

exports[`lineEllipseArcIntersections > should return multiple intersections of a line and an arc 1`] = `
[
  [
    -2.675224520712844,
    -2.675224520712844,
  ],
  [
    -1.004775479287156,
    -1.004775479287156,
  ],
]
`;
